#167573212 Implement Login Feature

henryade opened the pull request:
#### What does this PR do?
Implement login route

#### Description of Task proposed in this pull request?
- add login route
- test login route
- create user model, migration and seeds file
- add a `findUser` helper function
- add a password helper function
- fix houndCI issue

#### How should this be manually tested (Quality Assurance)?
- check into the branch by running `git checkout ft-login-167573212`
- create a `.env` file, copy the contents of the `.env.sample` file into it and update the fields and also your database details.
- run `npm db:rollback` to migrate and seed into the database.
- run `npm run start:dev` to start the server
- open [postman](https://www.getpostman.com/downloads/) and use seeded to the database to test e.g 
`POST http://localhost:9000/api/v1/sessions/create`
`{`
                `userLogin: 'JhayX'`
                `password: 'incorrect'`
`}`

#### What are the relevant pivotal tracker stories?
- Story ID:  [#167573212](https://www.pivotaltracker.com/story/show/167573212)

#### Any background context you want to add (Operations Impact)?
- run `npm install` to install all dependencies
- update the `.env` file with your database connection string
- run `npm run db:rollback` to create tables in the database and seed 

#### What I have learned working on this feature:
- I learnt how to mock server error test with sinon

#### Screenshots: 
Successful Login:
<img width="1434" alt="Screenshot 2019-07-30 at 6 56 22 PM" src="https://user-images.githubusercontent.com/47576037/62153214-fb0d5080-b2fb-11e9-8769-5bcec0e11d80.png">

Unsuccessful login:
<img width="1440" alt="Screenshot 2019-07-30 at 6 59 33 PM" src="https://user-images.githubusercontent.com/47576037/62154103-f9448c80-b2fd-11e9-83a1-dadf136a6dd7.png">

dinobi commented on .hound.yml:
```diff
@@ -1,6 +1,4 @@
 eslint:
   enabled: true
-  config_file: .eslintrc
+  config_file: .eslintrc.json
```
Why this change?

henryade replied:
To fix the hound probelm.

dinobi commented on .sequelizerc:
```diff
@@ -5,6 +5,6 @@ const path = require('path');
   'config': path.resolve('server/database/config', 'config.js'),
   'models-path': path.resolve('server/database/models'),
-  'seeds-path': path.resolve('server/database/seeds'),
+  'seeders-path': path.resolve('server/database/seeds'),
```
ditto

henryade replied:
the seeds script wasn't working. After researching, I found out I had to change the path. I also inform the database code owner.

dinobi commented on package.json:
```diff
@@ -12,11 +12,11 @@
     "db:migrate": "sequelize db:migrate",
     "db:migrate-undo": "sequelize db:migrate:undo:all",
-    "db:rollback": "npm run db:back && npm run db:roll && npm run db:seed",
+    "db:rollback": "npm run db:migrate-undo && npm run db:migrate && npm run db:seed",
```
ditto

henryade replied:
the db migrate script was formerly `db:roll` and you asked it to be changed in to `db: migrate` but the change wasn't reflected in the `db:rollback` script before merging. so for the script to work, I had to change it.

chialuka commented on server/controllers/user.js:
```diff
@@ -0,0 +1,50 @@
+import helpers from '../helpers';
```
Please, could you rename this file to upper camel case as it's default export is a class?

henryade replied:
Done

dinobi commented on server/database/migrations/20190729172658-create-user.js:
```diff
@@ -0,0 +1,64 @@
+module.exports = {
```
Why do we have user creation here

henryade replied:
I created it to test the login route with a user seeded into the database

dinobi commented on server/database/models/user.js:
```diff
@@ -0,0 +1,23 @@
+module.exports = (sequelize, DataTypes) => {
```
Why do we have user model here

henryade replied:
I created it to test the login route with a user seeded into the database

dinobi replied:
So why did you commit it?

> I committed it so that my teammates and travis can test the route. And also my test coverage will drop if Travis cannot test the new files. Should I will remove it?

dinobi commented on server/database/seeds/20190729181912-user.js:
```diff
@@ -0,0 +1,26 @@
+module.exports = {
```
Why this?

henryade replied:
same reason as above which is `I created it to test the login route with a user seeded into the database`

dinobi commented on test/users/user.test.js:
```diff
@@ -0,0 +1,91 @@
+import chai, { expect } from 'chai';
```
rename this to sessions.test.js

henryade replied:
okay

dinobi commented on server/helpers/dateHelper.js:
```diff
@@ -0,0 +1,14 @@
+  return expirationDate;
+};
+
+export default { expiryDate };
```
Is this result a datetime object?

henryade replied:
yes

dinobi commented on server/routes/index.js:
```diff
@@ -0,0 +1,8 @@
+
+const router = express.Router();
+
+router.use('/', user);
```
what is `user`?

henryade replied:
The user router signifying all routes pertaining to user auth. I initially named it auth.js but the team convention is to name the file user.

dinobi commented on server/routes/user.js:
```diff
@@ -0,0 +1,10 @@
+
+const router = express.Router();
+
+router.post('/users/login', create);
```
it should be `router.post('/sessions/create', Sessions.create)`
remove the destructing, cos we could possibly have other create methods, like User.create

henryade replied:
Done

dinobi commented on server/controllers/Sessions.js:
```diff
@@ -0,0 +1,78 @@
+// const { Session } = models;
+// const { findUser } = helpers;
+// const mobileDeviceIndicator = [
+//   'mobile',
```
What are these?

dinobi commented on server/controllers/Sessions.js:
```diff
@@ -0,0 +1,78 @@
+import bcrypt from 'bcryptjs';
+import {
+  findUser,
+  // generateToken,
```
This has been merged now

dinobi commented on server/controllers/Sessions.js:
```diff
@@ -0,0 +1,78 @@
+  static async create(req, res) {
+    try {
+      const { userLogin, password } = req.body;
+      if (!/\D/.test(userLogin)) {
```
I don't think we need this, just call `toString` on that variable if you must accept a string, regexes are computationally expensive

dinobi commented on server/controllers/Sessions.js:
```diff
@@ -0,0 +1,78 @@
+          message: 'user login must be a string'
+        });
+      }
+      const user = await findUser(userLogin);
```
How about we start thinking of models as objects that can hold properties, which is what they are. It would be nice to have this method that query a User table on the associated model object. I need all `rambo-devs` to reason in this light.

dinobi commented on server/controllers/Sessions.js:
```diff
@@ -0,0 +1,78 @@
+          message: 'incorrect username or password'
+        });
+      }
+      // const userAgent = req.headers['user-agent'];
```
please move this logic to a helper

dinobi commented on server/controllers/Sessions.js:
```diff
@@ -0,0 +1,78 @@
+      const { id, dataValues } = user;
+      // const expiresAt = expiryDate(devicePlatform);
+      // const token = generateToken(id);
+      // await Session.create({
```
update your PR

dinobi commented on server/controllers/Sessions.js:
```diff
@@ -0,0 +1,78 @@
+      //   devicePlatform
+      // });
+      delete dataValues.password;
+      return serverResponse(res, 200, { user: { ...dataValues } });
```
Is the token present in this response?, and also added to response header

dinobi reviewed (changes requested):
Please update your PR and test

henryade commented:
> Please update your PR and test

ayodejiAA reviewed (dismissed):
Looks okay. Good work

abdulfatai360 commented on server/docs/authors-haven-api.yml:
```diff
@@ -50,6 +50,53 @@ paths:
+                password:
+                  type: string
+                  minLength: 4
+                  example: incorrect
```
Ade, I am aware that there were no validation rules at the time you wrote this documentation. However, there is an adopted rule now. Could you please reflect them here in your documentation?

henryade committed:
feature-[167573212]: implement login feature
- added login route
- tested login route
- created user model, migration and seeds file
- added a findUser helper function
- added a password helper function that compares password

[finishes #167573212]

abdulfatai360 commented on server/controllers/Sessions.js:
```diff
@@ -0,0 +1,66 @@
+        });
+      }
+      const user = await findUser(userLogin);
+      let verifyPassword;
```
@henryade Could you please check if the request body is empty or either of the user login details is not present and return the appropriate response to the client?
